### PR TITLE
Give more specific error message on browser for nested policies

### LINKED ISSUE #4485: minio-gateway browser throws "Internal error" when one tries to add a nested bucket policy
## Expected Behavior

Minio browser UX throws a user-friendly error message explaining what went wrong.
minio-browser-policy

## Current Behavior

While, Minio gateway browser UI returns the following error.
minio-gateway-browser-policy

Minio gateway browser should also return a user-friendly like Minio browser UI does.

## Possible Solution

Not known yet
## Steps to Reproduce (for bugs)

1. Add bucket policy to nested paths e.g, `test/` and `test/prefix`.
2.
3.
4.

## Your Environment

* Version used (`minio version`):
```
Version: 2017-06-06T13:29:42Z
Release-Tag: DEVELOPMENT.2017-06-06T13-29-42Z
Commit-ID: a4d1ef1b62d2921a1711c189e48dc28700fbd50e

```
* Environment name and version (e.g. nginx 1.9.1):
* Server type and version:
* Operating System and version (`uname -a`):
* Link to your project:

### PR DESCRIPTION
**Pull request by poornas**
For Azure, nested policies are not implemented. Provide a more specific error message when user attempts to set nested policy. When all policies are removed for a bucket, the bucket policy also needs to be removed from gateway server.
## Motivation and Context

Fixes #4485 

## How Has This Been Tested?

Manually on Azure, S3 and server mode.
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by harshavardhana — DISMISSED** [review]

File: cmd/web-handlers.go [hunk 1]
```diff
@@ -1037,6 +1044,8 @@ func toWebAPIError(err error) APIError {
 		apiErrCode = ErrReadQuorum
 	case PolicyNesting:
 		apiErrCode = ErrPolicyNesting
+	case NotImplemented:
+		apiErrCode = ErrNotImplemented
```
Comment (donatello) [inline cmd/web-handlers.go]: For the original issue, the error message now shown is:
image

It says that functionality is not implemented, but the phrasing can be made more general. The missing functionality here is not really related to headers.

**Review by donatello — CHANGES_REQUESTED** [review]

File: cmd/web-handlers.go (hunk 1 above)
Comment (krishnasrinivas) [reply cmd/web-handlers.go]: I had suggested to use ErrNotImplemented, but @poornas you can use this as it gives more control on the error message:

```
krishna@escape:~/gocode/src/github.com/minio/minio$ git diff
diff --git a/cmd/web-handlers.go b/cmd/web-handlers.go
index bc4992e..29d25d8 100644
--- a/cmd/web-handlers.go
+++ b/cmd/web-handlers.go
@@ -1037,6 +1037,12 @@ func toWebAPIError(err error) APIError {
                apiErrCode = ErrReadQuorum
        case PolicyNesting:
                apiErrCode = ErrPolicyNesting
+       case NotImplemented:
+               return APIError{
+                       Code:           "NotImplemented",
+                       HTTPStatusCode: http.StatusBadRequest,
+                       Description:    "Functionality not implemented",
+               }
        default:
                // Log unexpected and unhandled errors.
                errorIf(err, errUnexpected.Error())
krishna@escape:~/gocode/src/github.com/minio/minio$ 
```

But now the code looks inconsistent as we return only in one case statement. So you can do this:
```
        // Convert error type to api error code.
        var apiErrCode APIErrorCode
        switch err.(type) {
        case StorageFull:
                return getAPIError(ErrStorageFull)
        case BucketNotFound:
                return getAPIError(ErrNoSuchBucket)
        case BucketExists:
                return getAPIError(ErrBucketAlreadyOwnedByYou)
        case BucketNameInvalid:
                return getAPIError(ErrInvalidBucketName)
        case BadDigest:
                return getAPIError(ErrBadDigest)
        case IncompleteBody:
                return getAPIError(ErrIncompleteBody)
        case ObjectExistsAsDirectory:
                return getAPIError(ErrObjectExistsAsDirectory)
        case ObjectNotFound:
                return getAPIError(ErrNoSuchKey)
        case ObjectNameInvalid:
                return getAPIError(ErrNoSuchKey)
        case InsufficientWriteQuorum:
                return getAPIError(ErrWriteQuorum)
        case InsufficientReadQuorum:
                return getAPIError(ErrReadQuorum)
        case PolicyNesting:
                return getAPIError(ErrPolicyNesting)
        case NotImplemented:
                return APIError{
                        Code:           "NotImplemented",
                        HTTPStatusCode: http.StatusBadRequest,
                        Description:    "Functionality not implemented",
                }
        }
        // Log unexpected and unhandled errors.
        errorIf(err, errUnexpected.Error())
        return APIError{
                Code:           "InternalError",
                HTTPStatusCode: http.StatusInternalServerError,
                Description:    err.Error(),
        }
```

Comment (poornas) [reply cmd/web-handlers.go]: updated the PR with suggestions

**Review by krishnasrinivas — APPROVED** [review]

**Review by donatello — APPROVED** [review]

File: cmd/web-handlers.go [hunk 2]
```diff
@@ -1011,39 +1018,46 @@ func toWebAPIError(err error) APIError {
...
-		apiErrCode = ErrReadQuorum
+		return getAPIError(ErrReadQuorum)
 	case PolicyNesting:
-		apiErrCode = ErrPolicyNesting
-	default:
-		// Log unexpected and unhandled errors.
-		errorIf(err, errUnexpected.Error())
-		apiErrCode = ErrInternalError
+		return getAPIError(ErrPolicyNesting)
+	case NotImplemented:
```
Comment (harshavardhana) [inline cmd/web-handlers.go:1046]: unit test for toWebAPIError

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**